x-pack/filebeat/input/gcs/job.go

```diff
@@ -184,53 +166,99 @@ func (j *job) readJsonAndPublish(ctx context.Context, r io.Reader, id string) er
 		}
 	}
 
-	var offset, relativeOffset int64
-	// uncompressed files use the client to directly set the offset, this
-	// in turn causes the offset to reset to 0 for the new stream, hence why
-	// we need to keep relative offsets to keep track of the actual offset
-	if !j.isCompressed && !j.isRootArray {
-		relativeOffset = j.offset
-	}
-	dec.UseNumber()
 	for dec.More() && ctx.Err() == nil {
 		var item json.RawMessage
-		offset = dec.InputOffset()
+		offset := dec.InputOffset()
 		if err = dec.Decode(&item); err != nil {
 			return fmt.Errorf("failed to decode json: %w", err)
 		}
-		// manually seek offset only if file is compressed or if root element is an array
-		if (j.isCompressed || j.isRootArray) && offset < j.offset {
+
+		// if expand_event_list_from_field is set, then split the event list
+		if j.src.ExpandEventListFromField != "" {
+			if err := j.splitEventList(j.src.ExpandEventListFromField, item, offset, j.hash, id); err != nil {
+				return err
+			}
 			continue
 		}
+
 		var parsedData []mapstr.M
 		if j.src.ParseJSON {
 			parsedData, err = decodeJSON(bytes.NewReader(item))
 			if err != nil {
 				j.log.Errorw("job encountered an error", "gcs.jobId", id, "error", err)
 			}
 		}
-		evt := j.createEvent(item, parsedData, offset+relativeOffset)
-		// updates the offset after reading the file
-		// this avoids duplicates for the last read when resuming operation
-		offset = dec.InputOffset()
-		// locks while data is being saved and published to avoid concurrent map read/writes
-		var (
-			done func()
-			cp   *Checkpoint
-		)
+		evt := j.createEvent(item, parsedData, offset)
 		if !dec.More() {
-			// if this is the last object, then peform a complete state save
-			cp, done = j.state.saveForTx(j.object.Name, j.object.Updated)
+			// if this is the last object, then perform a complete state save
+			cp, done := j.state.saveForTx(j.object.Name, j.object.Updated)
+			if err := j.publisher.Publish(evt, cp); err != nil {
+				j.log.Errorw("job encountered an error", "gcs.jobId", id, "error", err)
```

Comment:
A more detailed error message may be.. This is particularly when `publishing` , so something which tells that?